### PR TITLE
add ansible based operator

### PR DESCRIPTION
**Pull request by fmuyassarov**
Introduce an operator for life-cycle-management of NRI plugins. Users are expected to interact with the operator through a PluginConfig namespaced Custom Resource, which is structured as follows:

```yaml
apiVersion: config.nri/v1alpha1
kind: NriPluginDeployment
metadata:
  labels:
    app.kubernetes.io/name: nriplugindeployment
    app.kubernetes.io/instance: nriplugindeployment-sample
    app.kubernetes.io/part-of: nri-plugins-operator
    app.kubernetes.io/managed-by: kustomize
    app.kubernetes.io/created-by: nri-plugins-operator
  name: nriplugindeployment-sample
  namespace: kube-system
spec:
  pluginName: topology-aware
  pluginVersion: "v0.2.3"
  state: present
  values:
    nri:
      patchRuntimeConfig: true
```

- **`spec.pluginName`:** Indicates the preferred plugin to be installed. This field is an enum and can have values such as topology-aware, balloons, memtierd, memory-qos, or sgx-epc. The list is expected to grow as new plugins are introduced. The plugin name is intentionally immutable. To deploy a different plugin in the same namespace, either delete the Custom Resource (CR) or uninstall the existing plugin by setting `spec.state` to 'absent.' Multiple plugins cannot be installed in the same namespace.

- **`spec.pluginVersion`:** Specifies the version of the plugin. If not indicated, it defaults to the latest version. The plugin version is mutable, and updating it will uninstall the current version before installing the updated one.

- **`spec.state`:** Determines whether to install or uninstall the plugin. This field is an enum with `present` for installation and 'absent' for uninstallation. The state is mutable.

- **`spec.values`:** Allows users to provide custom values for manipulating Helm chart values, including plugin configuration data. This field is immutable, requiring users to recreate the CR to pass new values.

- **`status subresource`:** Tracks the basic state of the resource and includes basic messages in case the operator encounters issues.

To install, for example, the topology-aware plugin, users should create an instance of CR with `spec.[]` populated. To uninstall, either set `spec.state` to `absent` or delete the CR instance.

The Plugin Helm chart is installed in the same namespace as the CR instance.

### CODE REVIEW

**Comment by fmuyassarov** [comment]
@klihub @marquiz @askervin @kad PTAL. There are two more 3 commits I'm planning to push, for testing + image build, docs and loading values.yaml into cluster. But otherwise it can be already tested out.

File: config/samples/apps_v1alpha1_pluginconfig.yaml [hunk 1]
```diff
@@ -13,4 +13,4 @@ spec:
   pluginName: topology-aware
   pluginVersion: v0.2.3
   state: present
-  # pluginValuesFile: "/path/to/values.yaml"
+  # valuesFile: /tmp/value.yaml
```
Comment (klihub) [inline config/samples/apps_v1alpha1_pluginconfig.yaml]: I am probably missing a large part of the context here. Is this file expected to be found on the node where the operator ends up being scheduled to or on the node where the user is instantiating the operator on ?

Comment (fmuyassarov) [reply config/samples/apps_v1alpha1_pluginconfig.yaml]: I'm exploring possible options of loading local file into k8s and so far ConfigMap seems an option to go with but there are some drawbacks too. I will update once I'm satisfied with my current approach that is not pushed here yet.

File: config/crd/bases/config.nri_pluginconfigs.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            description: 'Kind is a string value representing the REST resource this
+              object represents. Servers may infer this from the endpoint the client
+              submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
+            type: string
+          metadata:
+            type: object
+          spec:
+            description: Spec defines the intended state of PluginConfig
+            properties:
+              pluginName:
```
Comment (askervin) [inline config/crd/bases/config.nri_pluginconfigs.yaml]: Is the master plan to create multiple PluginConfigs in order to get multiple NRI plugins installed on nodes, or is the plan to allow specifying multiple plugins, versions and their target nodes in the same PluginConfig? Obviously it's a perfectly valid use case to run `balloons` on some nodes, `topology-aware` on all the other nodes, and `memory-qos` on all nodes, for instance. Let's make sure that the design will enable getting there at some point.

**Review by askervin — COMMENTED** [review]
Very nice, Feruz! 

I see this operator logically as a layer on top of the built-in daemonset "operator". While the latter manages the life-cycle of a single NRI plugin in the cluster, the goal of this operator is to help managing a set of NRI plugins and their configurations.

Following this analogy, my first note would be on naming. As most of the CRDs are more or less configurations, I wouldn't include "config" in the name of the CRD nor the domain (DaemonSet is not a DaemonSetConfig either). And I think that "plugin" is too generic for a name of a kind, as there can be so many plugins everywhere in kubernetes. Could the CRD be called `NriPluginSets` or just `NriPlugins`?

File: config/samples/apps_v1alpha1_pluginconfig.yaml (hunk 1 above)
Comment (klihub) [reply config/samples/apps_v1alpha1_pluginconfig.yaml]: Yes, I also suspect that a volume-mounted inline ConfigMap looks like the only reasonable option for `valuesFile`.

**Comment by klihub** [comment]
> Very nice, Feruz!
> 
> I see this operator logically as a layer on top of the built-in daemonset "operator". While the latter manages the life-cycle of a single NRI plugin in the cluster, the goal of this operator is to help managing a set of NRI plugins and their configurations.
> 
> Following this analogy, my first note would be on naming. As most of the CRDs are more or less configurations, I wouldn't include "config" in the name of the CRD nor the domain (DaemonSet is not a DaemonSetConfig either). And I think that "plugin" is too generic for a name of a kind, as there can be so many plugins everywhere in kubernetes. Could the CRD be called `NriPluginSets` or just `NriPlugins`?

Maybe we could call it `PluginDeployment[.config.nri]` ?

**Comment by fmuyassarov** [comment]
> Maybe we could call it `PluginDeployment[.config.nri]` ?

`PluginDeployment` sounds good to me and it describes what it does.
@askervin WDYT?

File: config/crd/bases/config.nri_pluginconfigs.yaml (hunk 2 above)
Comment (fmuyassarov) [reply config/crd/bases/config.nri_pluginconfigs.yaml]: I designed it like one `PluginConfig` is for installing one NRI plugin on a given namespace. We could extend the CRD to have a list of plugins but then it means they will get deployed on the same namespace and I think that's not a good situation to be in?

**Comment by askervin** [comment]
> > Very nice, Feruz!
> > I see this operator logically as a layer on top of the built-in daemonset "operator". While the latter manages the life-cycle of a single NRI plugin in the cluster, the goal of this operator is to help managing a set of NRI plugins and their configurations.
> > Following this analogy, my first note would be on naming. As most of the CRDs are more or less configurations, I wouldn't include "config" in the name of the CRD nor the domain (DaemonSet is not a DaemonSetConfig either). And I think that "plugin" is too generic for a name of a kind, as there can be so many plugins everywhere in kubernetes. Could the CRD be called `NriPluginSets` or just `NriPlugins`?
> 
> Maybe we could call it `PluginDeployment[.config.nri]` ?

Or even `NriPluginDeployment`? I'm thinking of it how it shows in `kubectl api-resources` or `kubectl get crds`, and would like the plugin family (nri) to be included in the name due to "plugin" being so generic term. Analogously to having `csidrivers` instead of just `drivers`.

(PluginDeployment is already much better than PluginConfig, I can live with that, too...)

**Comment by fmuyassarov** [comment]
> kubectl api-resource

`NriPluginDeployment` is also fine with me and I can have maybe `npd` as short name.

File: config/samples/apps_v1alpha1_pluginconfig.yaml (hunk 1 above)
Comment (fmuyassarov) [reply config/samples/apps_v1alpha1_pluginconfig.yaml]: @klihub I have now removed the option to pass a file with custom values and instead made it possible to pass it directly via spec.

**Comment by klihub** [comment]
@fmuyassarov Is the current full commit history relevant enough to keep ? At a first glance there seems to be at least a few that look like good candidates for squashing. For instance these look like something you might want to omit from the initial history:
- rename group name to config.nri from apps.nri.io
-  CRD: rename CRD to be NriPluginDeployment
- clean up unused boiler plate code
- Revise RBAC rules for clarity and consistency
- CRD: allow passing helm values directly from CR spec
- Makefile: cleanup auto-generated comments

**Comment by fmuyassarov** [comment]
Sure. I can squash them. What do you think functionality wise?

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+BUNDLE_DEFAULT_CHANNEL := --default-channel=$(DEFAULT_CHANNEL)
+endif
+BUNDLE_METADATA_OPTS ?= $(BUNDLE_CHANNELS) $(BUNDLE_DEFAULT_CHANNEL)
+
+# IMAGE_TAG_BASE defines the docker.io namespace and part of the image name for remote images.
+# This variable is used to construct full image tags for bundle and catalog images.
+#
+# For example, running 'make bundle-build bundle-push catalog-build catalog-push' will build and push both
+# nri.io/nri-plugins-operator-bundle:$VERSION and nri.io/nri-plugins-operator-catalog:$VERSION.
+IMAGE_TAG_BASE ?= nri.io/nri-plugins-operator
```
Comment (klihub) [inline Makefile]: Is `nri.io` the namespace we'll use here ?

Comment (klihub) [reply Makefile]: Does it need to be consistent with `$IMG`, which uses `ghcr.io/containers` ?

File: PROJECT [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+plugins:
+  manifests.sdk.operatorframework.io/v2: {}
+  scorecard.sdk.operatorframework.io/v2: {}
+projectName: nri-plugins-operator
+resources:
+- api:
+    crdVersion: v1
+    namespaced: true
+  group: config.nri
+  kind: NriPluginDeployment
```
Comment (klihub) [inline PROJECT:17]: nitpick: since NRI is a TLA, should this be `NRIPluginDeployment` ?

**Comment by klihub** [comment]
> Sure. I can squash them. What do you think functionality wise?

I think it looks great functionality wise.

File: Makefile (hunk 3 above)
Comment (fmuyassarov) [reply Makefile]: fixed

**Comment by fmuyassarov** [comment]
@klihub 
I have now squashed some of the commits and improved others. PTAL and let me know if you find still some room for the improvement.

File: README.md [hunk 5]
```diff
@@ -1 +1,70 @@
...
+- `spec.values`: Allows users to provide custom values for manipulating Helm chart values. This field is immutable, 
+  requiring users to recreate the object to pass new values.
+- `spec.status`: Tracks the basic state of the resource and includes basic messages in case the operator encounters 
+  issues while reconciling the object.
+
+## Operator installation
+
+Build the operator image and push it to some registry
+```shell
+make docker-build docker-push IMG="ghcr.io/containers/nri-plugins-operator:unstable"
```
Comment (klihub) [inline README.md]: nitpick: that image location is not necessarily the best choice as an example here.

File: PROJECT [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+- ansible.sdk.operatorframework.io/v1
+plugins:
+  manifests.sdk.operatorframework.io/v2: {}
+  scorecard.sdk.operatorframework.io/v2: {}
+projectName: nri-plugins-operator
+resources:
+- api:
+    crdVersion: v1
+    namespaced: true
+  domain: nri.io
```
Comment (klihub) [inline PROJECT]: Is this now intentionally `nri.io` ?

File: PROJECT [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  group: apps
+  kind: PluginConfig
```
Comment (klihub) [inline PROJECT]: Is `PluginConfig` correct here ?

File: config/crd/bases/apps.nri.io_pluginconfigs.yaml [hunk 8]
```diff
@@ -0,0 +1,57 @@
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: pluginconfigs.apps.nri.io
+spec:
+  group: apps.nri.io
+  names:
```
Comment (klihub) [inline config/crd/bases/apps.nri.io_pluginconfigs.yaml]: Ditto here, do we use `nri.io` for these ?

**Comment by klihub** [comment]
> @klihub I have now squashed some of the commits and improved others. PTAL and let me know if you find still some room for the improvement.

Probably a matter of taste, but I'm still wondering if you want to keep the initial commit history like that. It had me sidetracked for a while as I was trying to re-review this commit by commit and it took me a few notes to realize that I'm reviewing outdated things that further commits down the line change.

Basically, as someone who is entirely unfamiliar with the operator framework, I see a few things here being added:
- CRD which acts as 'configuration/the user interface' for the operator
- YAML-codified operator logic to configure/code the actual operator framework/logic
- RBAC rules for the operator
- Dockerfile for building the operator image
- Makefile as the developer user interface
- docs describing what this operator is all about

If I understand the commits correctly, (pre-)initial versions of all of these are added in the first commit then modified in subsequent commits. You have 7 commits modifying the CRD, 1 modifying Dockerfile, 1 modifying the RBAC rules, 1 allowing ansible config customization, 1 to propagate and use the CR namespace for the installed plugin namespace, 1 altering comments in the Makefile, and 1 adding documentation.

Again, as I am unfamiliar with the operator framework and it's SDK, it is not clear to me if this choice for the commit series is meaningful/intuitive as opposed to, for instance, one commit per category listed above, or one commit for the whole shebang and another for docs...

You are the most familiar with operator framework so ultimately whatever you see fit is fine with me. Just wanted to give my n00b-perspective about reviewing this.

File: README.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    tolerations:
+      - key: "node-role.kubernetes.io/control-plane"
+        operator: "Exists"
+        effect: "NoSchedule"
+    affinity:
+      nodeAffinity:
+        requiredDuringSchedulingIgnoredDuringExecution:
+          nodeSelectorTerms:
+            - matchExpressions:
+                - key: topology.kubernetes.io/disk
```
Comment (askervin) [inline README.md:33]: Can we come up with an example of a node selection criteria that where the topology aware policy would be more obviously great choice. For instance: there is more than one NUMA node, or there are more than 3 CPU cores, or something like that.

If that is not easy to formulate with commonly available tools, then let's forget this. I'm anyway happy about the shape of this yaml.

Comment (fmuyassarov) [reply README.md:33]: I think that's a good idea but I don't have immediate idea to be honest how to depict it here. Perhaps I will take care of it in the next iteration as a future improvement and for now we maybe focus on getting the basic operator in place.

**Comment by fmuyassarov** [comment]
> You are the most familiar with operator framework so ultimately whatever you see fit is fine with me. Just wanted to give my n00b-perspective about reviewing this.

@klihub Sorry for the mess in the commit history. I must admit I was doing changes randomly, whatever was coming to my mind. I would probably be in the same situation as you when following these changes. I don't know how to make it easier to review but now what I did is I reduced the commits up to three where one is adding the operator code, the second for the CRD improvements and the last docs. I have tried to keep CRD related change notes as commit description. I know it is still hard to follow but I don't know how to make it better unfortunately. I can do some POC and code walk-through if that helps reviewing.

File: PROJECT (hunk 4 above)
Comment (fmuyassarov) [reply PROJECT:17]: as discussed during the call, let's keep it for now as it is and we can revisit it if we find some best practices that suggests using capitalized letters for the abbreviations.

File: README.md (hunk 9 above)
Comment (askervin) [reply README.md:33]: Yep, that's fine @fmuyassarov!

**Comment by klihub** [comment]
> > You are the most familiar with operator framework so ultimately whatever you see fit is fine with me. Just wanted to give my n00b-perspective about reviewing this.
> 
> @klihub Sorry for the mess in the commit history. I must admit I was doing changes randomly, whatever was coming to my mind. I would probably be in the same situation as you when following these changes.

No problemo. I don't think it was really a mess, it was just kept intact as it happened commit by commit... which is not necessarily what you want for a first real commit. 

>  I don't know how to make it easier to review

What I meant by the review difficulties is that it was not possible/easy to review commit by commit, because things added in the first commit were altered in significant ways in later commits. For instance things like 'domain' and 'Kind' were not the final ones in the first commit, which I think is also later reflected in some of the other files getting renamed in subsequent commits...

> but now what I did is I reduced the commits up to three where one is adding the operator code, the second for the CRD improvements and the last docs. I have tried to keep CRD related change notes as commit description. I know it is still hard to follow but I don't know how to make it better unfortunately. I can do some POC and code walk-through if that helps reviewing.

I have one question about the current version. Despite what the inline comments say about the generated PROJECT file, we do edit it manually here. Is that ok ? I guess it should be as long as the final version is consistent with the rest of the repo content and we made sure here that it is, right ?

Anyway, I think we shouldn't spend too much time on iterating the commit history if you are happy with the current one (and editing the PROJECT file is done so that any related tooling won't go berzerk).

**Comment by fmuyassarov** [comment]
> Anyway, I think we shouldn't spend too much time on iterating the commit history if you are happy with the current one (and editing the PROJECT file is done so that any related tooling won't go berzerk).

I did manual update because I didn't want to re-create a new CRD or generate a new project from scratch only for updating CRD. But anyways I had to create a new second CRD with the final name and group name and then copies content of the PROJECT file and pasted it on this PR's PROJECT. In other words, it should be safe even though it is highly discouraged.

**Review by klihub — APPROVED** [review]
> I did manual update because I didn't want to re-create a new CRD or generate a new project from scratch only for updating CRD. But anyways I had to create a new second CRD with the final name and group name and then copies content of the PROJECT file and pasted it on this PR's PROJECT. In other words, it should be safe even though it is highly discouraged.

Then LGTM.

**Comment by askervin** [comment]
I gave this a shot following the README: deployed the operator with `make deploy` and, in order to get operator working, ran: `kubectl apply -k ./config/samples`.

Deploying topology-aware NRI plugin seems to fail. You can find log output below. Error message looks pretty nice in the NriPluginDeployment's status:

```
kubectl describe nriplugindeployment nriplugindeployment-sample  -n kube-system
...
Status:
  Conditions:
    Last Transition Time:  2023-11-29T14:40:28Z
    Message:
    Reason:
    Status:                False
    Type:                  Successful
    Ansible Result:
      Changed:             0
      Completion:          2023-11-29T14:53:07.252657
      Failures:            1
      Ok:                  0
      Skipped:             0
    Last Transition Time:  2023-11-29T14:53:07Z
    Message:               Failure when executing Helm command. Exited 1.
stdout:
stderr: Error: looks like "https://containers.github.io/nri-plugins" is not a valid chart repository or cannot be reached: Get "https://containers.github.io/nri-plugins/index.yaml": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

    Reason:                Failed
    Status:                False
    Type:                  Failure
    Last Transition Time:  2023-11-29T14:53:07Z
    Message:               Running reconciliation
    Reason:                Running
    Status:                True
    Type:                  Running
...
```


This probably won't bring in any extra information, but I'll attach logs just in case:
```
kubectl logs -n nri-plugins-operator-system nri-plugins-operator-controller-manager-7784fdff-6zdgl
...
{"level":"info","ts":"2023-11-29T14:38:25Z","logger":"logging_event_handler","msg":"[playbook taskstart]","name":"nriplugindeployment-sample","namespace":"kube-system","gvk":"config.nri/v1alpha1, Kind=NriPluginDeployment","event_type":"playbook_on_task_start","job":"5665556735360545279","EventData.Name":"plugin-deployment : Add nri-plugins helm chart repo"}
{"level":"error","ts":"2023-11-29T14:40:28Z","logger":"logging_event_handler","msg":"","name":"nriplugindeployment-sample","namespace":"kube-system","gvk":"config.nri/v1alpha1, Kind=NriPluginDeployment","event_type":"runner_on_failed","job":"5665556735360545279","EventData.Task":"Add nri-plugins helm chart repo","EventData.TaskArgs":"","EventData.FailedTaskPath":"/opt/ansible/roles/plugin-deployment/tasks/main.yml:2","error":"[playbook task failed]","stacktrace":"github.com/operator-f
ramework/ansible-operator-plugins/internal/ansible/events.loggingEventHandler.Handle\n\tansible-operator-plugins/internal/ansible/events/log_events.go:111"}

--------------------------- Ansible Task StdOut -------------------------------

 TASK [Add nri-plugins helm chart repo] ********************************

TASK [plugin-deployment : Add nri-plugins helm chart repo] *********************
task path: /opt/ansible/roles/plugin-deployment/tasks/main.yml:2
fatal: [localhost]: FAILED! => {"changed": false, "command": "/usr/local/bin/helm repo add nri-plugins https://containers.github.io/nri-plugins", "msg": "Failure when executing Helm command. Exited 1.\nstdout: \nstderr: Error: looks like \"https://containers.github.io/nri-plugins\" is not a valid chart repository or cannot be reached: Get \"https://containers.github.io/nri-plugins/index.yaml\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)\n", "stderr": "Error: looks like \"https://containers.github.io/nri-plugins\" is not a valid chart repository or cannot be reached: Get \"https://containers.github.io/nri-plugins/index.yaml\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)\n", "stderr_lines": ["Error: looks like \"https://containers.github.io/nri-plugins\" is not a valid chart repository or cannot be reached: Get \"https://containers.github.io/nri-plugins/index.yaml\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"], "stdout": "", "stdout_lines": []}

-------------------------------------------------------------------------------
{"level":"error","ts":"2023-11-29T14:40:28Z","logger":"runner","msg":"\u001b[0;34mansible-playbook [core 2.15.3]\u001b[0m\r\n\u001b[0;34m  config file = /etc/ansible/ansible.cfg\u001b[0m\r\n\u001b[0;34m  configured module search path = ['/usr/share/ansible/openshift']\u001b[0m\r\n\u001b[0;34m  ansible python module location  = /usr/local/lib/python3.9/site-packages/ansible\u001b[0m\r\n\u001b[0;34m  ansible collection location = /opt/ansible/.ansible/collections:/usr/share/ansible/collections\u001b[0m\r\n\u001b[0;34m  executable location = /usr/local/bin/ansible-playbook\u001b[0m\r\n\u001b[0;34m  python version = 3.9.16 (main, May 31 2023, 12:21:
58) [GCC 8.5.0 20210514 (Red Hat 8.5.0-18)] (/usr/bin/python3)\u001b[0m\r\n\u001b[0;34m  jinja version = 3.1.2\u001b[0m\r\n\u001b[0;34m  libyaml = True\u001b[0m\r\n\u001b[0;34mUsing /etc/ansible/ansible.cfg as config file\u001b[0m\r\n\u001b[0;34mSkipping callback 'awx_display', as we already have a stdout callback.\u001b[0m\
n\u001b[0;34mSkipping callback 'default', as we already have a stdout callback.\u001b[0m\n\u001b[0;34mSkipping callback 'minimal', as we already have a stdout callback.\u001b[0m\n\u001b[0;34mSkipping callback 'oneline', as we already have a stdout callback.\u001b[0m\n\r\nPLAYBOOK: plugin-deployment.yml **********************
*
```

**Comment by askervin** [comment]
This may be a proxy issue. I wouldn't consider it blocking, but probably nice to get working at some point for validation in environments behind proxy.

Here's what I got when installed helm on the same single node cluster host where I ran the test above:

```
root@n4c16-debian-sid-containerd:/home/debian/helm/linux-amd64# ./helm repo add nri-plugins https://containers.github.io/nri-plugins
"nri-plugins" has been added to your repositories
root@n4c16-debian-sid-containerd:/home/debian/helm/linux-amd64# ./helm repo rm nri-plugins
"nri-plugins" has been removed from your repositories
root@n4c16-debian-sid-containerd:/home/debian/helm/linux-amd64# unset http_proxy https_proxy HTTP_PROXY HTTPS_PROXY
debian-sid-containerd:/home/debian/helm/linux-amd64# ./helm repo add nri-plugins https://containers.github.io/nri-plugins
Error: looks like "https://containers.github.io/nri-plugins" is not a valid chart repository or cannot be reached: Get "https://containers.github.io/nri-plugins/index.yaml": dial tcp 185.199.110.153:443: i/o timeout (Client.Timeout exceeded while awaiting headers)
```

**Review by askervin — APPROVED** [review]
With the note on issues behind proxies, I'm fine merging this and fixing the issue later.

LGTM

**Comment by fmuyassarov** [comment]
@askervin I just did another round of end to end testing and could not reproduce the same issue. But as you pointed out, it seems due to proxy settings in your environment and I didn't have any proxies setting.
